bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -210,12 +210,14 @@ void ReplicaImp::sendRaw(char *m, NodeIdType dest, uint16_t type, MsgSize size)
 IncomingMsg ReplicaImp::recvMsg() {
   while (true) {
     auto msg = incomingMsgsStorage.pop(timersResolution);
-    if (msg.tag != IncomingMsg::INVALID) {
-      return msg;
-    }
+
     // TODO(GG): make sure that we don't check timers too often
     // (i.e. much faster than timersResolution)
     timersScheduler.evaluate();
+
```

Comment:
This change makes it so that timers are evaluated on every message which we don't want. While the code isn't necessarily clear as to what `IncomingMsg::Invalid` means, we only want to check when the message is Invalid. Essentially that return code means that the thread local queue is empty and it happens when a condVar timer expires and their are no more messages to pop. You can see that this behavior follows the original intent of the code when modified to use smart pointers: https://github.com/vmware/concord-bft/commit/8583ded17c29e627244f199cbfa5f58737b4f9b5#diff-48324f7e49aa046d298836b4ccc2a318L230-L239

Please revert this change and also add a message relating to what I just said for bonus points. I'm glad you actually did this because it points out a place where the intent isn't clear. Perhaps we should also change `IncomingMsg::Invalid` to `IncomingMsg::QueueEmpty` or something.

Reply:
OK. I'll do that. Can you help me understand one thing first, though?  I added that particular change because in practice I saw little or no logging without it.  The commit message related to this change was:

    Prevent timer starvation in bftengine main loop under load.
    
    Before this commit, ReplicaImp::recvMsg() tries to obtain an incoming
    message and, if that fails, attempts to fire any pending timers.  If
    messages are arriving faster than the replica can process them, this means
    that timers never fire.  I actually observed this starvation in simpleTest
    running on my laptop.
    
    This commit fixes the problem by ensuring that timers are evaluated every
    time through the main loop rather than just when no messages are available.

Is there some other change that should have been made instead to avoid the timer starvation? Or is this a sign that I invoked the tests wrong, etc.?

Reply:
I take it back Ben. I missed the commit message by not looking at it directly and only looking at the GitHub comment. I didn't realize how intentional it was. That was my fault.  I think this is a good mechanism to avoid the timer starvation problem. Please disregard my prior comment.

Reply:
Theoretically we could use some sort of counter in addition to prevent timer starvation, and not have to check on every message, but I don't see a ton of harm here either.